### PR TITLE
parquet datasource

### PR DESCRIPTION
**Pull request by Risers**

### CODE REVIEW

File: storage/parquet/datasource.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	int64s     []int64
+	int96s     []parquet.Int96
+	float32s   []float32
+	float64s   []float64
+	byteArrays [][]byte
+
+	values interface{}
+
+	dLevels []uint16
+	rLevels []uint16
```
Comment (cube2222) [inline storage/parquet/datasource.go]: Please add a comment about what those variables are, and how they differ.

File: storage/parquet/datasource.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+	cr *parquet.ColumnChunkReader
+	rg int
+	n  int
+	i  int
+	vi int
```
Comment (cube2222) [inline storage/parquet/datasource.go]: Please make those variable names more descriptive.

File: storage/parquet/datasource.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+func (ds *DataSource) Get(ctx context.Context, variables octosql.Variables) (execution.RecordStream, error) {
+	file, err := parquet.OpenFile(ds.path)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't open file")
+	}
+	columns := file.Schema.Columns()
+	n := len(columns)
+	for _, col := range columns {
+		if col.MaxR() != 0 {
+			return nil, fmt.Errorf("column '%s' has repeated elements", col)
```
Comment (cube2222) [inline storage/parquet/datasource.go]: Could you please elaborate what this means? (best in a comment.)

File: storage/parquet/datasource.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+	m := make(map[string]interface{})
+	for i, it := range rs.columnIterators {
+		s, err := it.Next()
+		if err != nil {
+			if err == io.EOF {
+				rs.isDone = true
+				err = rs.file.Close()
+				if err != nil {
+					return nil, errors.Wrap(err, "Couldn't close underlying file")
```
Comment (cube2222) [inline storage/parquet/datasource.go]: Start error wrap message with lowercase please.

File: storage/parquet/datasource.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	}
+
+	return nil
+}
+
+func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
+	if rs.isDone {
+		err := rs.file.Close()
+		if err != nil {
+			return nil, errors.Wrap(err, "Couldn't close underlying file")
```
Comment (cube2222) [inline storage/parquet/datasource.go]: Start error wrap message with lowercase please.

File: storage/parquet/datasource.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	columnIterators []*ColStrIter
+	columns         []parquet.Column
+	isDone          bool
+	alias           string
+}
+
+func (rs *RecordStream) Close() error {
+	err := rs.file.Close()
+	if err != nil {
+		return errors.Wrap(err, "Couldn't close underlying file")
```
Comment (cube2222) [inline storage/parquet/datasource.go]: Start error wrap message with lowercase please.

File: storage/parquet/datasource.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+	byteArrays [][]byte
+
+	values interface{}
+
+	dLevels []uint16 // definition levels
+	rLevels []uint16 // repetition levels
+
+	cr                  *parquet.ColumnChunkReader
+	rg                  int // row groups
+	n                   int // number of loaded elements
```
Comment (cube2222) [inline storage/parquet/datasource.go]: There are so many variables, please don't abbreviate them. (just use rowGroups or similar)

File: storage/parquet/datasource.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+		err := rs.file.Close()
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't close underlying file")
+		}
+		return nil, execution.ErrEndOfStream
+	}
+
+	m := make(map[string]interface{})
+
+	rs.isDone = true
```
Comment (woojtek) [inline storage/parquet/datasource.go:268]: don't operate on `rs.isDone` directly - instead use some temporary variable and afterwards set `if <variable> { rs.isDone = true}`

File: storage/parquet/datasource.go [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+			return nil, errors.Wrap(err, "couldn't close underlying file")
+		}
+		return nil, execution.ErrEndOfStream
+	}
+
+	aliasedRecord := make(map[octosql.VariableName]octosql.Value)
+	for k, v := range m {
+		if str, ok := v.(string); ok {
+			parsed, err := time.Parse(time.RFC3339, str)
+			if err == nil {
```
Comment (woojtek) [inline storage/parquet/datasource.go:296]: You should check whether `err != nil` instead, try to rewrite this way

Comment (Risers) [reply storage/parquet/datasource.go:291]: this code is copied from the json data source and I think with err != nil check this code will be less readable

Comment (woojtek) [reply storage/parquet/datasource.go:291]: sure

File: storage/parquet/datasource.go [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+					s = it.float32s[it.dataIterator]
+				case parquetformat.Type_DOUBLE:
+					s = it.float64s[it.dataIterator]
+				case parquetformat.Type_BYTE_ARRAY, parquetformat.Type_FIXED_LEN_BYTE_ARRAY:
+					s = it.byteArrays[it.dataIterator]
+				default:
+					panic("unknown type")
+				}
+			} else {
+				switch it.col.Type() {
```
[the remark this answers is not in the document]
Comment (Risers) [reply storage/parquet/datasource.go:154]: marked as resolved same as above

**Pull request merged**